File: lr/src/questions/q_0102_binary_tree_level_order_traversal.rs
use std::{cell::RefCell, collections::VecDeque, rc::Rc};

use crate::utils::TreeNode;

pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    if root.is_none() {
        return vec![];
    }
    let mut result = vec![];
    let mut q = VecDeque::new();
    q.push_back(root);

    while !q.is_empty() {
        let len = q.len();
        let mut level = vec![];
        for _ in 0..len {
            let node = q.pop_front().unwrap();
            if let Some(node) = node {
                level.push(node.as_ref().borrow().val);
                q.push_back(node.as_ref().borrow().left.clone());
                q.push_back(node.as_ref().borrow().right.clone());
            }
        }
        if !level.is_empty() {
            result.push(level);
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use crate::tree;

    use super::*;

    #[test]
    fn test_level_order() {
        assert_eq!(
            level_order(tree!(3, 9, 20, None, None, 15, 7)),
            vec![vec![3], vec![9, 20], vec![15, 7]]
        );
    }
}

File: lg/src/questions/q_0102_binary_tree_level_order_traversal.go
package questions

import "github.com/hi-rustin/lg/src/utils"

func levelOrder(root *utils.TreeNode) [][]int {
	if root == nil {
		return nil
	}

	result := make([][]int, 0)
	queue := make([]*utils.TreeNode, 0)
	queue = append(queue, root)
	for len(queue) > 0 {
		queueLen := len(queue)
		level := make([]int, 0, queueLen)
		for queueLen > 0 {
			node := queue[0]
			level = append(level, node.Val)
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			queueLen -= 1
		}
		result = append(result, level)
	}

	return result
}
